### PR TITLE
RUM-4134 Remove moshi dependency from trace module

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/sampling/SpanSamplingRules.java [hunk 1]
```diff
@@ -1,209 +1,263 @@
...
+            log.error("Couldn't parse Span Sampling Rules from JSON: {}", json, ex);
+        }
+        return result;
     }
 
-    private static void logRuleError(JsonRule rule, String error) {
-      log.error("Skipping invalid Span Sampling Rule: {} - {}", rule, error);
+    public static SpanSamplingRules deserializeFile(String jsonFile) {
+        SpanSamplingRules result = EMPTY;
+        try (com.google.gson.stream.JsonReader reader = new JsonReader(new FileReader(jsonFile))) {
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/main/java/com/datadog/trace/common/sampling/SpanSamplingRules.java:49]: it is already in the import

```suggestion
        try (JsonReader reader = new JsonReader(new FileReader(jsonFile))) {
```

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesFileTest.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+package com.datadog.trace.common.sampling
+
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+
+internal class SpanSamplingRulesFileTest : SpanSamplingRulesTest() {
+
+    @TempDir
+    lateinit var tempDir: File
+    override fun deserializeRules(jsonRules: String): SpanSamplingRules {
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesFileTest.kt:16]:
```suggestion
    lateinit var tempDir: File

    override fun deserializeRules(jsonRules: String): SpanSamplingRules {
```

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesTest.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    @ParameterizedTest
+    @ValueSource(strings = ["-0.1", "-11", "1.2", "100", "\"zero\"", "\"\""])
+    fun `Skip Span Sampling Rules with invalid sample_rate values`(rate: String) {
+        // Given
+        val rules =
+            deserializeRules(
+                """[
+            {"service": "usersvc", "name": "healthcheck", "sample_rate": $rate}
+        ]"""
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesTest.kt:104]: seems like formatting is a bit broken for raw strings in this file

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/TraceSamplingRulesTest.kt [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    @Test
+    fun `Keep only valid rules when invalid rules are present`() {
+        val rules = TraceSamplingRules.deserialize(
+            """[
+            {"service": "usersvc", "name": "healthcheck", "sample_rate": 0.5},
+            {"service": "usersvc", "name": "healthcheck", "sample_rate": 200}
+        ]"""
+        )
+        assertThat(rules.rules).hasSize(1)
```
Comment (0xnm) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/TraceSamplingRulesTest.kt:121]:
```suggestion
        // When
        val rules = TraceSamplingRules.deserialize(
            """[
            {"service": "usersvc", "name": "healthcheck", "sample_rate": 0.5},
            {"service": "usersvc", "name": "healthcheck", "sample_rate": 200}
        ]"""
        )
        
        // Then
        assertThat(rules.rules).hasSize(1)
```

**Review by 0xnm — DISMISSED** [review]
nice job!

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesTest.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
+package com.datadog.trace.common.sampling
```
Comment (jonathanmos) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/SpanSamplingRulesTest.kt:1]: minor: missing copyright

File: features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/TraceSamplingRulesTest.kt [hunk 6]
```diff
@@ -0,0 +1,123 @@
+package com.datadog.trace.common.sampling
```
Comment (jonathanmos) [inline features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/common/sampling/TraceSamplingRulesTest.kt:1]: minor: missing copyright

**Review by jonathanmos — APPROVED** [review]

**Review by xgouchet — APPROVED** [review]

**Review by 0xnm — APPROVED** [review]

**Pull request merged**